test/e2e/storage/provision.go

```diff
@@ -0,0 +1,112 @@
+package storage
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+
+	snapshotv1 "github.com/kubernetes-csi/external-snapshotter/client/v6/apis/volumesnapshot/v1"
+	snapshotclientset "github.com/kubernetes-csi/external-snapshotter/client/v6/clientset/versioned"
+	v1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2ekubectl "k8s.io/kubernetes/test/e2e/framework/kubectl"
+	"k8s.io/kubernetes/test/e2e/framework/pv"
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+)
+
+var _ = ginkgo.Describe("provision volumes with different volume modes from volume snapshot dataSource", func() {
+	var local struct {
+		snapshotClient *snapshotclientset.Clientset
+		sc             *storagev1.StorageClass
+		vsc            *snapshotv1.VolumeSnapshotClass
+	}
+	f := framework.NewDefaultFramework("pvcs-from-volume-snapshots")
+	ginkgo.BeforeEach(func() {
+		var err error
+		output, err := e2ekubectl.RunKubectl("kube-system", "describe", "deployment", "snapshot-controller")
+		if err != nil || !strings.Contains(output, "prevent-volume-mode-conversion=true") {
+			e2eskipper.Skipf("--prevent-volume-mode-conversion feature flag is not enabled in snapshot-controller. Skipping volume mode conversion tests.")
+		}
+		output, err = e2ekubectl.RunKubectl("default", "describe", "sts", "csi-hostpathplugin")
+		if err != nil || !strings.Contains(output, "prevent-volume-mode-conversion=true") {
+			e2eskipper.Skipf("--prevent-volume-mode-conversion feature flag is not enabled in csi-hostpathplugin. Skipping volume mode conversion tests.")
+		}
+
+		ginkgo.By("Creating CSI Hostpath driver Storage Class")
+		local.sc, err = createHostPathStorageClass(f.ClientSet)
+		framework.ExpectNoError(err)
+		local.snapshotClient, err = snapshotclientset.NewForConfig(f.ClientConfig())
+		framework.ExpectNoError(err)
+		ginkgo.By("Creating VolumeSnapshotClass")
+		local.vsc, err = createVolumeSnapshotClass(local.snapshotClient, snapshotv1.VolumeSnapshotContentDelete)
+		framework.ExpectNoError(err)
+	})
+	ginkgo.AfterEach(func() {
+		if local.snapshotClient != nil && local.vsc != nil {
+			ginkgo.By("Deleting VolumeSnapshotClass")
+			err := local.snapshotClient.SnapshotV1().VolumeSnapshotClasses().Delete(context.TODO(), local.vsc.Name, metav1.DeleteOptions{})
+			framework.ExpectNoError(err)
+		}
+		if local.sc != nil {
+			ginkgo.By("Deleting CSI Hostpath driver Storage Class")
+			err := f.ClientSet.StorageV1().StorageClasses().Delete(context.TODO(), local.sc.Name, metav1.DeleteOptions{})
+			framework.ExpectNoError(err)
+		}
+	})
+
+	// Attempt to create a PVC that alters the volume mode of the source volume snapshot.
+	// The PVC is expected to remain in the Pending phase
+	ginkgo.It("when the source volume mode is altered without permissions", func() {
```

Comment:
Can you add a test when there is permission?

Reply:
I planned to add that in a follow up PR to manage the size of this one.

Reply:
Ok